packages/gatsby-cli/src/util/normalize-windows-cwd.js

```diff
@@ -0,0 +1,59 @@
+const { platform, tmpdir } = require(`os`)
+const report = require(`../reporter`)
+
+/**
+ * This function ensures that the current working directory on Windows
+ * always has an uppercase drive letter (i.e., C: vs. c:).
+ *
+ * Why?
+ * 1. Different utils like "true-case-path", "normalize-path", "slash" treat Windows
+ * drive letter differently. "true-case-path" will uppercase, others usually don't care.
+ * As a result path normalization produces different results depending on current cwd (c: vs. C:)
+ * which manifests in weird bugs that are very hard to debug.
+ *
+ * We can't control community plugins or site code, so everything should be working
+ * even with a different set of libraries.
+ *
+ * Related: https://github.com/Profiscience/true-case-path/issues/3
+ *
+ * 2. Builds save some paths in a cache. If you run the first build from "c:" shell
+ * and then the next one from "C:" shell, you may get a bunch of webpack warnings
+ * because it expects module paths to be case-sensitive.
+ */
+module.exports = function normalizeWindowsCwd() {
+  if (platform() !== `win32`) {
+    return
+  }
+  const cwd = process.cwd()
+  const normalizedCwd = driveLetterToUpperCase(cwd)
+
+  if (cwd !== normalizedCwd) {
+    try {
+      // When cwd is "c:\dir" then command "cd C:\dir" won't do anything
+      // You have to change the dir twice to actually change the casing of the path
+      process.chdir(tmpdir())
+      process.chdir(normalizedCwd)
+    } catch {
+      // rollback
+      process.chdir(cwd)
+    }
+
+    if (normalizedCwd !== process.cwd()) {
+      report.warn(
+        report.stripIndent(`
+          Your working directory has a lowercase drive letter:
+            "${cwd}".
+          For stable cross-platform builds we recommend switching it to:
+            "${normalizedCwd}".
+        `)
```

Comment:
I wonder about value of this warning: so far our strategy for stable builds is to use relative paths with unix-style slashes in places that would affect build artifacts - for example: https://github.com/gatsbyjs/gatsby/blob/47d2a8d39b96e151036dcb0e01e3cbcf3be8d1b0/packages/gatsby/src/bootstrap/requires-writer.js#L169-L177

which should render drive letter case irrelevant for output?

Is there actual build differences if you use differently cased drive letter with your PR right now?

In case this still matters - can we provide actual commands for user to run to change cwd? (as user might need to switch to tmp dir as in code you have to change dir programatically)

Reply:
Oh, right. I guess it only happens in develop here:

https://github.com/gatsbyjs/gatsby/blob/47d2a8d39b96e151036dcb0e01e3cbcf3be8d1b0/packages/gatsby/src/bootstrap/requires-writer.js#L147-L161

But it all started from stable builds issue in https://github.com/gatsbyjs/gatsby/pull/15876 - so that's the initial source of confusion. I will rephrase:

 `For stable cross-platform builds we recommend switching it to` to `For consistent development experience` or do you have other suggestions?